sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousCoalesceRDD.scala

```diff
@@ -0,0 +1,108 @@
+  // This flag will be flipped on the executors to indicate that the threads processing
+  // partitions of the write-side RDD have been started. These will run indefinitely
+  // asynchronously as epochs of the coalesce RDD complete on the read side.
+  private[continuous] var writersInitialized: Boolean = false
+}
+
+/**
+ * RDD for continuous coalescing. Asynchronously writes all partitions of `prev` into a local
+ * continuous shuffle, and then reads them in the task thread using `reader`.
+ */
+class ContinuousCoalesceRDD(
+    context: SparkContext,
+    numPartitions: Int,
+    readerQueueSize: Int,
+    epochIntervalMs: Long,
+    readerEndpointName: String,
+    prev: RDD[InternalRow])
+  extends RDD[InternalRow](context, Nil) {
+
+  override def getPartitions: Array[Partition] = Array(ContinuousCoalesceRDDPartition(0))
+
+ val readerRDD = new ContinuousShuffleReadRDD(
+    sparkContext,
+    numPartitions,
+    readerQueueSize,
+    prev.getNumPartitions,
+    epochIntervalMs,
+    Seq(readerEndpointName))
+
+  private lazy val threadPool = ThreadUtils.newDaemonFixedThreadPool(
+    prev.getNumPartitions,
+    this.name)
+
+  override def compute(split: Partition, context: TaskContext): Iterator[InternalRow] = {
+    assert(split.index == 0)
+    // lazy initialize endpoint so writer can send to it
+    readerRDD.partitions(0).asInstanceOf[ContinuousShuffleReadPartition].endpoint
+
+    if (!split.asInstanceOf[ContinuousCoalesceRDDPartition].writersInitialized) {
+      val rpcEnv = SparkEnv.get.rpcEnv
+      val outputPartitioner = new HashPartitioner(1)
+      val endpointRefs = readerRDD.endpointNames.map { endpointName =>
+          rpcEnv.setupEndpointRef(rpcEnv.address, endpointName)
+      }
+
+      val runnables = prev.partitions.map { prevSplit =>
+        new Runnable() {
+          override def run(): Unit = {
+            TaskContext.setTaskContext(context)
+
+            val writer: ContinuousShuffleWriter = new RPCContinuousShuffleWriter(
+              prevSplit.index, outputPartitioner, endpointRefs.toArray)
+
+            EpochTracker.initializeCurrentEpoch(
+              context.getLocalProperty(ContinuousExecution.START_EPOCH_KEY).toLong)
+            while (!context.isInterrupted() && !context.isCompleted()) {
+              writer.write(prev.compute(prevSplit, context).asInstanceOf[Iterator[UnsafeRow]])
+              // Note that current epoch is a non-inheritable thread local, so each writer thread
+              // can properly increment its own epoch without affecting the main task thread.
+              EpochTracker.incrementCurrentEpoch()
+            }
+          }
+        }
+      }
+
+      context.addTaskCompletionListener { ctx =>
+        threadPool.shutdownNow()
+      }
+
+      split.asInstanceOf[ContinuousCoalesceRDDPartition].writersInitialized = true
+
+      runnables.foreach(threadPool.execute)
+    }
+
+    readerRDD.compute(readerRDD.partitions(split.index), context)
```

Comment:
The `writer.write` and `readerRDD.compute()` is going to execute as a separate tasks (but concurrently since there are no stage boundaries) correct?

Reply:
If its the same task, then do we need the RPC mechanism to pass the rows around ?

Reply:
There is a queue inside the `ContinuousShuffleReadRDD` that is buffering all the records that are being sent out by the `RPCContinuousShuffleWriter`. And the compute function is returning data from that queue.

As I commented above, we dont really need the ContinuousShuffleReadRDD, just the ContinuousShuffleReader

Reply:
Agree. Also the 2*numShuffle partition threads here looks like an overhead. Maybe ok for now but the CoalesceRDD iterator could just iterate over the parent RDD partitions tracking the epochs, returning the rows and terminating when the epoch is received from all its parents.

Reply:
Yeah, it could be made more efficient. Part of the goal here is to ensure that the shuffling does indeed work end-to-end, so we can work on both the shuffle framework and distributed repartitioning in parallel.